### ISSUE TITLE
GPT as a Language Model

### ISSUE THREAD

**Issue by mdasadul** [huggingface/pytorch-pretrained-BERT#473]
I am interested to use GPT as Language Model to assign Language modeling score (Perplexity score) of a sentence.  Here is what I am using 
```import torch
import math
from pytorch_pretrained_bert import OpenAIGPTTokenizer, OpenAIGPTModel, OpenAIGPTLMHeadModel
# Load pre-trained model (weights)
model = OpenAIGPTLMHeadModel.from_pretrained('openai-gpt')
model.eval()
# Load pre-trained model tokenizer (vocabulary)
tokenizer = OpenAIGPTTokenizer.from_pretrained('openai-gpt')

def score(sentence):
    tokenize_input = tokenizer.tokenize(sentence)
    tensor_input = torch.tensor([tokenizer.convert_tokens_to_ids(tokenize_input)])
    loss=model(tensor_input, lm_labels=tensor_input)
    return math.exp(loss)


a=['there is a book on the desk',
                'there is a plane on the desk',
                        'there is a book in the desk']
print([score(i) for i in a])
21.31652459381952, 61.45907380241148, 26.24923942649312
```
Is it the right way to score a sentence ?

**Comment by thomwolf** [comment]
Looks good to me. It's perplexity so lower is better.
You can do a `math.exp(loss.item())` to be a little cleaner.

**Comment by thomwolf** [comment]
Oh no wait, you need to compare to the shifted inputs:
`loss=model(tensor_input[:-1], lm_labels=tensor_input[1:])`
It's a causal model, it predicts the next token given the previous ones.

**Comment by mdasadul** [comment]
I can see inside the ```class OpenAIGPTLMHeadModel(OpenAIGPTPreTrainedModel)```  this shifting is happening 
```        if lm_labels is not None:
            # Shift so that tokens < n predict n
            shift_logits = lm_logits[:, :-1].contiguous()
            shift_labels = lm_labels[:, 1:].contiguous()```

Do I still need to use ```loss=model(tensor_input[:-1], lm_labels=tensor_input[1:])```?

**Comment by thomwolf** [comment]
Oh you are right, this has been added now with #404.
So the way you are doing looks fine to me.
Shifting the logics inside the model can a bit dangerous for the people who are used to train a causal model the usual way, I'll add a mention in the README.

**Comment by mdasadul** [comment]
Thanks for your quick response. I can see there is a minor bug when I am trying to predict with a sentence which has one word. You can re create the error by using my above code.
```sentence='Learn'
score(sentence)```

**Comment by thomwolf** [comment]
Unfortunately, given the way the model is trained (without using a token indicating the beginning of a sentence), I would say it does not make sense to try to get a score for a sentence with only one word.

**Comment by asad1996172** [comment]
How can we use this to get the probability of a particular token? So, for instance, let's say we have the following sentence. 
"He was going home"
and we want to get the probability of "home" given the context "he was going"

**Comment by rohuns** [comment]
@mdasadul aren't those scores too large to be a log probability?

**Comment by apappu97** [comment]
@thomwolf If the shifting of the lm_labels matrix isn't necessary (before passing into the model's forward method) because of the internal logit shifting, should the preprocess code for finetuning GPT1 in RocStories be changed? At [https://github.com/huggingface/pytorch-pretrained-BERT/blob/master/examples/run_openai_gpt.py#L86](url), I believe the continuations are shifted over in lm_labels one relative to input_ids.

**Comment by thomwolf** [comment]
Oh yes, of course! Do you want to submit a PR on that? Otherwise I'll take of it later.

**Comment by apappu97** [comment]
For sure! Will look at it soon.

**Comment by UpcominTerminator** [comment]
@thomwolf  Hey how can I give my own checkpoint files to the model while loading

**Comment by ssdavidson** [comment]
When I attempt this, with the exact same code and the exact same sentences - I get some wildly different numbers. Anyone have any idea why this might be?

`import math
from pytorch_pretrained_bert import OpenAIGPTTokenizer, OpenAIGPTModel, OpenAIGPTLMHeadModel
import torch
# Load pre-trained model (weights)
model = OpenAIGPTLMHeadModel.from_pretrained('openai-gpt')
model.eval()
# Load pre-trained model tokenizer (vocabulary)
tokenizer = OpenAIGPTTokenizer.from_pretrained('openai-gpt')

def score(sentence):
    tokenize_input = tokenizer.tokenize(sentence)
    tensor_input = torch.tensor([tokenizer.convert_tokens_to_ids(tokenize_input)])
    loss=model(tensor_input, lm_labels=tensor_input)
    return math.exp(loss)


a=['there is a book on the desk',
                'there is a plane on the desk',
                        'there is a book in the desk']
print([score(i) for i in a])
[12792.86796714989, 7661.653323354917, 6540.017210278059]
`

**Comment by ssdavidson** [comment]
I am using the exact same code as posted by the OP, but I am getting wildly different output. Does anyone have an idea as to why this may be happening?

'''
import math
from pytorch_pretrained_bert import OpenAIGPTTokenizer, OpenAIGPTModel, OpenAIGPTLMHeadModel
import torch
# Load pre-trained model (weights)
model = OpenAIGPTLMHeadModel.from_pretrained('openai-gpt')
model.eval()
# Load pre-trained model tokenizer (vocabulary)
tokenizer = OpenAIGPTTokenizer.from_pretrained('openai-gpt')

def score(sentence):
    tokenize_input = tokenizer.tokenize(sentence)
    tensor_input = torch.tensor([tokenizer.convert_tokens_to_ids(tokenize_input)])
    loss=model(tensor_input, lm_labels=tensor_input)
    return math.exp(loss)


a=['there is a book on the desk',
                'there is a plane on the desk',
                        'there is a book in the desk']
print([score(i) for i in a])
'''
And my output: 
'''
[12792.86796714989, 7661.653323354917, 6540.017210278059]
'''

The only difference I can see in my code is a call to import torch. Other than that it is verbatim.

Any help would be greatly appreciated.

**Comment by Lilo2305** [comment]
Is this score normalized on sentence lenght? And if not, what do I need to change to normalize it?

**Comment by Namrata96** [comment]
I'm confused whether the right way to calculate the perplexity for GPT2 is what the OP has done or as per the documentation https://huggingface.co/transformers/perplexity.html? Or  both are equivalent for some value of the stride?
